### PR TITLE
HPCC-18406 Use global and local ids to track related transactions

### PR DESCRIPTION
**Pull request by afishbeck**
In order to create a giant graph of every system call related to a
global product order/request, an external system may pass a globalId
and a "callerId" to an HPCC component.

GlobalIds tie together the entire customer order/request across
all components.

CallerIds tie together all the calls made from a particular transaction
node.

When globalIds are recieved we should:

1. Log the global and caller ids that are received.
2. Create our own local id representing the specific request on
   the current node.
2. When making any SOAPCALL or HTTPCALL automatically pass along the
   globalId and pass our localId as the outgoing "callerId".

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by AttilaVamos** [comment]
@afishbeck It seems your changes alter the test output, but the keyfile didn't changed and that's why httpcall_multiheader.ecl failed.

File: common/thorhelper/thorsoapcall.cpp [hunk 1]
```diff
@@ -1524,6 +1524,12 @@ class CWSCAsyncFor : implements IWSCAsyncFor, public CInterface, public CAsyncFo
         if (!httpHeaderBlockContainsHeader(httpheaders, ACCEPT_ENCODING))
             request.appendf("%s: gzip, deflate\r\n", ACCEPT_ENCODING);
 #endif
+        if (!isEmptyString(master->logctx.queryGlobalId()))
+        {
+            request.append("RELX-Global-Id: ").append(master->logctx.queryGlobalId()).append("\r\n");
```
Comment (richardkchapman) [inline common/thorhelper/thorsoapcall.cpp]: This feels rather RELX specific for OSS code. Is there a reason the Global ID would only be useful to RELX?

Comment (afishbeck) [reply common/thorhelper/thorsoapcall.cpp]: It's more that all RISK components will be using the same headers.  I.e. web applications, scapps.  Now is the time to set the standard, we could do HPCC, but not sure what would fit for all of RISK (and perhaps eventually beyond).

Comment (richardkchapman) [reply common/thorhelper/thorsoapcall.cpp]: What about users of HPCC outside of RELX ? Why would they want to use a global id called "RELX-Global-Id" ?

Comment (afishbeck) [reply common/thorhelper/thorsoapcall.cpp]: Perhaps something even more generic.  I thought just Global-ID might not be unique enough.

**Comment by afishbeck** [comment]
@richardkchapman note the inclusion of libbase58 code which is MIT license.  I'm not sure what the current practice is.  For now simply adding it was simplest.  MIT licesnse is very flexible.  base58 was chosen for the IDs (and already in use in RISK india) because it's human readable.  I.e. a customer could read the number to a customer support person.

File: common/thorhelper/thorsoapcall.cpp (hunk 1 above)
Comment (afishbeck) [reply common/thorhelper/thorsoapcall.cpp]: I'll bring up a more generic name with the Global ID team.

**Comment by afishbeck** [comment]
@AttilaVamos right, thanks, I'll add the update for the httpcall_multiheader.ecl key.

File: common/thorhelper/thorsoapcall.cpp (hunk 1 above)
Comment (richardkchapman) [reply common/thorhelper/thorsoapcall.cpp]: Alternatively, make the name configurable.

**Comment by AttilaVamos** [comment]
@afishbeck back to you.

File: common/thorhelper/thorsoapcall.cpp (hunk 1 above)
Comment (afishbeck) [reply common/thorhelper/thorsoapcall.cpp]: Yes, I'll do that.

**Comment by richardkchapman** [comment]
Needs rebasing for a clean merge

**Comment by afishbeck** [comment]
@richardkchapman updated to make it more generic for external users, and made it configurable so we could match the internal HTTP header name.

Also rebased.

**Pull request merged**